### PR TITLE
CORE-4870 - Remove normalisation of CPK metadata in the DB

### PR DESCRIPTION
**Pull request by driessamyn**
Removing normalisation of cpk metadata in the DB and replace with storing JSON encoded AVRO object.

Related API change: https://github.com/corda/corda-api/pull/379

### CODE REVIEW

File: components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt [hunk 1]
```diff
@@ -243,49 +239,16 @@ class DatabaseChunkPersistence(private val entityManagerFactory: EntityManagerFa
...
             em.merge(CpkDataEntity(cpkChecksum, Files.readAllBytes(it.path!!)))
 
-            val cpkMetadataEntity = CpkMetadataEntity(
+            val cpkMetadataEntity = CpkEntity(
                 cpi = cpiMetadataEntity,
-                cpkFileChecksum = cpkChecksum,
+                // TODO - do we need this double/triple conversion or could metadata always exist in the avro schema
+                //  with CpkMetadata just being a proxy for it?
+                // TODO - format version
+                metadata = CpkMetadataEntity(cpkChecksum, "1", it.metadata.toJsonAvro()),
```
Comment (driessamyn) [inline components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt]: FYI @josephzunigadaly - not sure if this fits in with the plans for format version.

File: components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseChunkPersistenceTest.kt [hunk 2]
```diff
@@ -433,4 +442,33 @@ internal class DatabaseChunkPersistenceTest {
             persistence.persistMetadataAndCpks(cpi, "test.cpi", checksum, UUID.randomUUID().toString(), "123456")
         }
     }
+
+    @Test
```
Comment (driessamyn) [inline components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseChunkPersistenceTest.kt:446]: @vkolomeyko FYI

I think we need integration tests like these for the reset functionality, as I think relying on the E2E tests is too slow in terms of feedback and too cumbersome in terms of debugging issues.

These tests are fairly lightweight as they can run against the in-memory DB.

**Review by josephzunigadaly — DISMISSED** [review]

**Review by josephzunigadaly — DISMISSED** [review]

File: gradle.properties [hunk 3]
```diff
@@ -39,7 +39,7 @@ commonsTextVersion = 1.9
 # Corda API libs revision (change in 4th digit indicates a breaking change)
 # Change to 5.0.0.xx-SNAPSHOT to pick up maven local published copy
 #cordaApiVersion=5.0.0.XX-SNAPSHOT
-cordaApiVersion=5.0.0.106-beta+
+cordaApiVersion=5.0.0.107-alpha-1653500339542
```
Comment (barrylapthorn) [inline gradle.properties]: ?

Comment (driessamyn) [reply gradle.properties]: I want to prove this works before merging the api repo. That's the process we should follow. As soon as the api repo is merged, I can update this to 107-beta+

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkEntity.kt [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    @Version
+    @Column(name = "entity_version", nullable = false)
+    var entityVersion: Int = 0
+    // this TS is managed on the DB itself
+    @Column(name = "insert_ts", insertable = false, updatable = false)
+    val insertTimestamp: Instant? = null
+}
+
+/** The composite primary key for a [CpkEntity]. */
+@Embeddable
```
Comment (kyriathar) [inline libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkEntity.kt:67]: I think `@Embeddable` is not needed here.

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkEntity.kt [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+    // NOTE: not mapping to CPK data as it would be a potentially bad idea to fetch all the CPK data.
+    //  generally one-by-one loading is going to be better for the files.
+    //  However, metadata is always loaded.
+    @Id
+    @OneToOne(cascade = [CascadeType.MERGE])
+    @JoinColumns(
+        JoinColumn(name = "cpk_file_checksum", referencedColumnName = "file_checksum", insertable = false, updatable = false)
+    )
+    val metadata: CpkMetadataEntity,
```
Comment (kyriathar) [inline libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkEntity.kt:51]: I think this looks a bit problematic in that in order for `CpkEntity` to be persisted, means `CpkDataEntity` for the corresponding `cpk_file_checksum` needs to be saved first. However, saving the `CpkDataEntity` first is not forced by the entities relationship. Maybe we should mention that on a comment?

**Review by josephzunigadaly — APPROVED** [review]

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkEntity.kt [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+ * @property cpkFileName for the CPK as defined in the CPI.
+ *                      NOTE: it is possible to have CPKs know by different filenames in case they
+ *                      where packaged up in different CPIs.
+ *                      This property should only be used for information, not as a key.
+ * @property insertTimestamp when the CPK Entity was inserted
+ */
+@Entity
+@Table(name = "cpi_cpk", schema = DbSchema.CONFIG)
+@IdClass(CpkEntityKey::class)
+data class CpkEntity(
```
Comment (kyriathar) [inline libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkEntity.kt:33]: Not too sure but , maybe for future work, could it be done to even more simplify here by only having `CpkMetadataEntity` and then that also gets the filename from `cpi_cpk` table? Then `CpiMetadataEntity` has list of `CpkMetadataEntity`.

**Review by kyriathar — COMMENTED** [review]
Late review after merge. Minor things.

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkEntity.kt (hunk 4 above)
Comment (driessamyn) [reply libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkEntity.kt:67]: I thought they were needed for the composite keys?

Comment (kyriathar) [reply libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkEntity.kt:67]: I think it is not needed for `@IdClass`.

Comment (driessamyn) [reply libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkEntity.kt:67]: I just tried and then you need a default constructor, which isn't great.

**Pull request merged**